spark/src/main/scala/org/apache/hadoop/fs/versioned/VersionedFileSystem.scala

```diff
@@ -0,0 +1,138 @@
+
+    import scala.collection.JavaConverters._
+
+    mappings = conf.iterator.asScala
+      .filter(_.getKey.startsWith(VersionPaths.KEY_PREFIX))
+      .map(_.getValue)
+      .toList
+      .map { v =>
+        (v, v + "/" + version)
+      }
+      .toMap
+
+    // versioned filesystem URIs take the form `versioned:URI`, e.g:
+    //    versioned:s3://my-bucket
+    //    versioned:file://tmp
+
+    // When initialising the versioned filesystem, we need to extract the
+    // URI for the base filesystem, which is accessible under the scheme
+    // specific part of the URI.
+    baseURI = new URI(baseFSScheme, null, name.getSchemeSpecificPart, null, null)
+    baseFS = FileSystem.get(baseURI, conf)
+    baseFS.initialize(baseURI, conf)
+  }
+
+  override def getUri: URI =
+    baseURI
+
+  override def open(f: Path, bufferSize: Int): FSDataInputStream =
+    baseFS.open(setScheme(baseURI.getScheme)(f), bufferSize)
+
+  override def create(
+      f: Path,
+      permission: FsPermission,
+      overwrite: Boolean,
+      bufferSize: Int,
+      replication: Short,
+      blockSize: Long,
+      progress: Progressable): FSDataOutputStream =
+    baseFS.create(baseVersionedPath(f), permission, overwrite, bufferSize, replication, blockSize, progress)
+
+  override def append(f: Path, bufferSize: Int, progress: Progressable): FSDataOutputStream =
+    baseFS.append(baseVersionedPath(f), bufferSize, progress)
+
+  override def rename(src: Path, dst: Path): Boolean =
+    baseFS.rename(baseVersionedPath(src), baseVersionedPath(dst))
+
+  override def delete(f: Path, recursive: Boolean): Boolean =
+    baseFS.delete(baseVersionedPath(f), recursive)
+
+  override def listStatus(f: Path): Array[FileStatus] =
+    baseFS.listStatus(setScheme(baseURI.getScheme)(f)) map { fileStatus =>
+      new FileStatus(
+        fileStatus.getLen,
+        fileStatus.isDirectory,
+        fileStatus.getReplication.toInt,
+        fileStatus.getBlockSize,
+        fileStatus.getModificationTime,
+        versionedPath(fileStatus.getPath)
+      )
+    } filter { fileStatus =>
+      fileStatus.getPath.toUri.getSchemeSpecificPart.contains(version) ||
+      fileStatus.isDirectory
+    }
+
+  override def setWorkingDirectory(new_dir: Path): Unit =
+    baseFS.setWorkingDirectory(baseVersionedPath(new_dir))
+
+  override def getWorkingDirectory: Path =
+    setScheme(baseURI.getScheme)(baseFS.getWorkingDirectory)
+
+  override def mkdirs(f: Path, permission: FsPermission): Boolean =
+    baseFS.mkdirs(baseVersionedPath(f), permission)
+
+  override def getFileStatus(f: Path): FileStatus = {
+    val fileStatus = baseFS.getFileStatus(setScheme(baseURI.getScheme)(f))
+
+    new FileStatus(
+      fileStatus.getLen,
+      fileStatus.isDirectory,
+      fileStatus.getReplication.toInt,
+      fileStatus.getBlockSize,
+      fileStatus.getModificationTime,
+      versionedPath(fileStatus.getPath)
+    )
```

Comment:
There are a few places where `FileStatus` objects are converted so it'd be worth pulling this out into a separate method. See for example the `swizzleFileStatus` method here: https://github.com/apache/hive/blob/master/shims/common/src/main/java/org/apache/hadoop/fs/ProxyFileSystem.java#L61 (but don't copy the name! 😁 )

Reply:
Good point. Done